Mage/src/main/java/mage/abilities/effects/ContinuousEffectsList.java

```diff
@@ -23,14 +25,20 @@ public class ContinuousEffectsList<T extends ContinuousEffect> extends ArrayList
 
     // the effectAbilityMap holds for each effect all abilities that are connected (used) with this effect
     private final Map<UUID, Set<Ability>> effectAbilityMap = new HashMap<>();
+    // Keep track of which effects have hints
+    private final List<WeakReference<T>> hints = new ArrayList<>();
 
     public ContinuousEffectsList() {
     }
 
     public ContinuousEffectsList(final ContinuousEffectsList<T> effects) {
         this.ensureCapacity(effects.size());
         for (ContinuousEffect cost : effects) {
-            this.add((T) cost.copy());
+            T copy = (T) cost.copy();
+            this.add(copy);
+            if (copy.hasHint()) {
+                hints.add(new WeakReference<>(copy));
```

Comment:
No need to keep filtered list in the same place and all that potentially buggy sync code. Process all effects on needs. It’s fast.

Reply:
Okay, I was worried it'd be too slow with too many permanents and effects, but I'll change it. Should be easy enough.